instrumentation/okhttp/okhttp-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/okhttp/v3_0/OkHttp3InstrumentationModule.java

```diff
@@ -42,16 +46,65 @@ public OkHttp3InstrumentationModule() {
 
     @Override
     public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
-      return singletonMap(
+      Map<ElementMatcher.Junction<MethodDescription>, String> transformers = new HashMap<>();
+      transformers.put(
           isConstructor().and(takesArgument(0, named("okhttp3.OkHttpClient$Builder"))),
-          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3Advice");
+          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3ClientConstructorAdvice");
+      transformers.put(
+          isMethod().and(named("newBuilder").and(returns(named("okhttp3.OkHttpClient$Builder")))),
+          OkHttp3InstrumentationModule.class.getName() + "$OkHttp3ClientNewBuilderAdvice");
+      return transformers;
     }
   }
 
-  public static class OkHttp3Advice {
+  // This advice makes two guarantees:
+  // 1) The state of the builder (specifically interceptor list) is the same before and after the
+  //    OkHttpClient constructor invocation
+  // 2) The interceptor list of the created OkHttpClient has exactly one instance of the tracing
+  //    interceptor and it is in the end (assuming no other instrumentations)
+  public static class OkHttp3ClientConstructorAdvice {
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void addTracingInterceptor(@Advice.Argument(0) OkHttpClient.Builder builder) {
+    public static void addTracingInterceptor(
+        @Advice.Argument(0) OkHttpClient.Builder builder,
+        @Advice.Local("otelOriginalInterceptors") List<Interceptor> originalInterceptors) {
+
+      if (builder.interceptors().contains(OkHttp3Interceptors.TRACING_INTERCEPTOR)) {
+        // Potential corner case - the tracing interceptor may be in the builder due to the builder
```

Comment:
What's the difference vs just calling return without the state management?

Reply:
A possible case is that the tracing interceptor may be in the middle, therefore it captures a "request" that may not be the actual request at all due to some later interceptor deciding to do something different.

Reply:
I don't think it can be in the middle if we always add it in the constructor (unless already present which is only when constructing a builder from another client, where it again couldn't be in the middle).

Reply:
It is possible - the third test in `OkHttp3Test` produces this scenario.